tendermint/src/time.rs

```diff
@@ -171,14 +299,47 @@ mod tests {
             "2021-01-07T20:26:12.078268Z",
             "2021-01-07T20:26:13.08074100Z",
             "2021-01-07T20:26:15.079663000Z",
-        ];
+        ]
+        .into_iter()
+        .map(String::from)
+        .collect();
 
-        for input in DATES {
-            let initial_time: Time = input.parse().unwrap();
-            let encoded_time = serde_json::to_value(&initial_time).unwrap();
-            let decoded_time = serde_json::from_value(encoded_time.clone()).unwrap();
+        select(strs)
+    }
+
+    proptest! {
+        #[test]
+        fn can_parse_rfc3339_timestamps(stamp in arb_rfc3339_timestamp()) {
+            prop_assert!(stamp.parse::<Time>().is_ok())
+        }
+
+        #[test]
+        fn serde_from_value_is_the_inverse_of_to_value_within_reasonable_time_range(
+            datetime in arb_datetime()
+        ) {
+            // If `from_value` is the inverse of `to_value`, then it will always
+            // map the JSON `encoded_time` to back to the inital `time`.
+            let time: Time = datetime.into();
+            let json_encoded_time = serde_json::to_value(&time).unwrap();
+            let decoded_time: Time = serde_json::from_value(json_encoded_time).unwrap();
+            prop_assert_eq!(time, decoded_time);
+        }
 
-            assert_eq!(initial_time, decoded_time);
+        #[test]
+        fn serde_of_rfc3339_timestamps_is_safe(
+            stamp in prop_oneof![
+                arb_rfc3339_timestamp(),
+                particular_rfc3339_timestamps(),
+            ]
+        ) {
+            // ser/de of rfc3339 timestamps is safe if it never panics.
+            // This differes from the the inverse test in that we are testing on
+            // arbitrarily generated textual timestamps, rather than times in a
+            // range. Tho we do incidentally test the inversion as well.
+            let time: Time = stamp.parse().unwrap();
+            let json_encoded_time = serde_json::to_value(&time).unwrap();
+            let decoded_time: Time = serde_json::from_value(json_encoded_time).unwrap();
+            prop_assert_eq!(time, decoded_time);
```

Comment:
This test ends up duplicating elements of the previous two, but it is ultimately testing a different property than the other two, as explained in the comment.

I'm not sure how I feel about this overlap. I could remove the `can_parse` test, since that is fully subsumed by this property, but I think we should maintain two separate tests, one to check for arbitrarily generated `DateTime` structs (as we do for `serde_from_value`) and one to check for arbitrarily generated time stamps (as we do here).

I"m open to other suggestions tho!